### DISCUSSION TITLE
Creating Pages For Tags (Contentful + Gatsby website)

### DISCUSSION THREAD

**Discussion by thedivtagguy** [gatsbyjs/gatsby#31881] [Help]
Each of my posts on Contentful has some associated categories with it. For example, one post might contain:
```
major: "HCD"
year: "1st Year"
tools: ["R", "Python", "Wordpress"]
```
On the website:
![image](https://user-images.githubusercontent.com/6006888/121716144-4fd7a000-cafd-11eb-9154-d5dffd8da9d9.png)

I am trying to create a page for each of these categories and tags. For example, if a user clicks on `Photoshop`, they should be taken to a page `tags/photoshop` and all posts containing that tag should be listed out. 

Fortunately, I was able to find [this guide](https://dennytek.com/blog/personal-site-with-gatsby-part-7/) to help me do this. However, the guide is not for Contentful data so I'm having a bit of trouble on how to do this. I have created the `tagsTemplate.jsx` and but I'm stuck at creating the actual pages.

My `gatsby-node.js` file looks like this:

```js
const path = require(`path`)
const _ = require('lodash');

exports.createSchemaCustomization = ({ actions }) => {
  const { createTypes } = actions
  const typeDefs = `
    type contentfulPortfolioDescriptionTextNode implements Node {
      description: String
      major: String
      author: String
      tools: [String]
      files: [ContentfulAsset]
      contact: String
    }
    type ContentfulPortfolio implements Node {
      description: contentfulPortfolioDescriptionTextNode
      gallery: [ContentfulAsset]
      id: ID!
      name: String!
      related: [ContentfulPortfolio]
      slug: String!
      major: String!
      files: [ContentfulAsset]
      author: String!
      tools: [String]!
      year: String!
      thumbnail: ContentfulAsset
      url: String
      contact: String
    }
  `
  createTypes(typeDefs)
}

exports.createPages = ({ graphql, actions }) => {
  const { createPage } = actions

  return new Promise((resolve, reject) => {
    graphql(`
      {
        portfolio: allContentfulPortfolio {
          nodes {
            slug
            tools
          }
        }
      }
    `).then(({ errors, data }) => {
      if (errors) {
        reject(errors)
      }

      if (data && data.portfolio) {
        const component = path.resolve("./src/templates/portfolio-item.jsx")
        data.portfolio.nodes.map(({ slug }) => {
          createPage({
            path: `/${slug}`,
            component,
            context: { slug },
          })
        })
      }
       
      const tools = data.portfolio.nodes.tools;
      const tagTemplate = path.resolve(`src/templates/tagsTemplate.js`);
      let tags = [];
      // Iterate through each post, putting all found tags into `tags`
      tags = tags.concat(tools);
       // Eliminate duplicate tags
      tags = _.uniq(tags);
      
       // Make tag pages
       tags.forEach(tag => {
        createPage({
          path: `/tags/${_.kebabCase(tag)}/`,
          component: tagTemplate,
          context: {
            tag,
          },
        });
      });
      resolve()
    })
  })
}
```

When I visit the page for one of the tags I know exists (like `photoshop`), I get a 404. What am I doing wrong and how can I fix it?

**Comment by axe312ger** [comment]
Hey Aman,

How do you connect to Contentful?

You probably want to use [gatsby-source-contentful](https://www.gatsbyjs.com/plugins/gatsby-source-contentful/). With that plugin, you don't have to create your own schema and your Contentful space data will be duplicated into your Gatsby instance. If you do that, your `createPages` looks fine on a first glimbse and you can remove `createSchemaCustomization `

As soon the page runs, and your tag pages do not exist, I would recommend you to check http://localhost:8000/___graphql. There you can dig into your GraphQL data and check what is there.

Hope that helps,
Benedikt
